File: console/console.fs
module console.console

open System

open core.sudoku
open core.hints

open format

// Things we may want to write
type consoleChar = 
    | CChar of char
    | CStr of string
    | ColouredChar of char * ConsoleColor
    | ColouredString of string * ConsoleColor
    | NL

let cs = CChar >> Seq.singleton

// Some predefined characters - for smaller grid
let defaultGridChars : gridChars<seq<consoleChar>> = 
    { h = cs '─'
      v = 
          { l = cs '│'
            m = cs '│'
            r = cs '│' }
      t = 
          { l = cs '┌'
            m = cs '┬'
            r = cs '┐' }
      m = 
          { l = cs '├'
            m = cs '┼'
            r = cs '┤' }
      b = 
          { l = cs '└'
            m = cs '┴'
            r = cs '┘' }
      n = Seq.singleton NL }

// Some predefined characters - for smaller grid
let defaultCandidateGridChars : candidateGridChars<seq<consoleChar>> = 
    { h = cs '═'
      hi = cs '─'
      v = 
          { l = cs '║'
            m = cs '║'
            r = cs '║' }
      vi = cs '│'
      t = 
          { mi = cs '╦'
            x = 
                { l = cs '╔'
                  m = cs '╦'
                  r = cs '╗' } }
      m = 
          { mi = cs '╬'
            x = 
                { l = cs '╠'
                  m = cs '╬'
                  r = cs '╣' } }
      mi = 
          { mi = cs '┼'
            x = 
                { l = cs '╠'
                  m = cs '╬'
                  r = cs '╣' } }
      b = 
          { mi = cs '╧'
            x = 
                { l = cs '╚'
                  m = cs '╩'
                  r = cs '╝' } }
      n = Seq.singleton NL }

// Change the console colour to write a string
let consoleWriteColor (value : 'a) consoleColour = 
    let foregroundColour = System.Console.ForegroundColor
    try 
        System.Console.ForegroundColor <- consoleColour
        System.Console.Write value
    finally
        System.Console.ForegroundColor <- foregroundColour

let drawConsoleChar (consoleChar : consoleChar) = 
    match consoleChar with
    | CChar c -> Console.Write c
    | CStr c -> Console.Write c
    | ColouredChar(c, consoleColour) -> consoleWriteColor c consoleColour
    | ColouredString(c, consoleColour) -> consoleWriteColor c consoleColour
    | NL -> Console.WriteLine ""

let drawDigitCellContents (firstDigit : digit option) (currentDigit : cellContents) = 
    match firstDigit, currentDigit with
    | Some s, _ -> ColouredString(s.ToString(), ConsoleColor.Blue)
    | None, BigNumber s -> ColouredString(s.ToString(), ConsoleColor.Red)
    | None, PencilMarks _ -> CChar '.'

let drawDigitCellContentAnnotations centreCandidate (candidate : digit) (firstDigit : digit option) 
    (currentDigit : cellContents) (annotationOpt : annotation option) = 
    let isCentre = centreCandidate = candidate

    match firstDigit, currentDigit with
    | Some _, _ when not isCentre -> CChar ' '
    | Some s, _ -> 
        match annotationOpt with
        | Some annotation when annotation.primaryHintHouse -> 
            ColouredString(s.ToString(), ConsoleColor.Cyan)
        | Some annotation when annotation.secondaryHintHouse -> 
            ColouredString(s.ToString(), ConsoleColor.DarkBlue)
        | _ -> ColouredString(s.ToString(), ConsoleColor.Blue)

    | None, BigNumber _ when not isCentre -> CChar ' '
    | None, BigNumber s -> 
        match annotationOpt with
        | Some annotation when annotation.primaryHintHouse -> 
            ColouredString(s.ToString(), ConsoleColor.Yellow)
        | Some annotation when annotation.secondaryHintHouse -> 
            ColouredString(s.ToString(), ConsoleColor.DarkRed)
        | _ -> ColouredString(s.ToString(), ConsoleColor.Red)
    | None, PencilMarks candidates -> 
        match annotationOpt with
        | Some annotation when annotation.setValue.IsSome && annotation.setValue.Value = candidate -> 
                ColouredString(candidate.ToString(), ConsoleColor.Red)
        | Some annotation when annotation.setValue.IsSome && Set.contains candidate candidates -> 
            ColouredString(candidate.ToString(), ConsoleColor.DarkYellow)
        | Some annotation when annotation.setValueReduction.IsSome && annotation.setValueReduction.Value = candidate && Set.contains candidate candidates -> 
                ColouredString(candidate.ToString(), ConsoleColor.DarkYellow)
        | Some annotation when Set.contains candidate annotation.reductions -> 
            ColouredString(candidate.ToString(), ConsoleColor.DarkYellow)
        | Some annotation when Set.contains candidate annotation.pointers -> 
            ColouredString(candidate.ToString(), ConsoleColor.Magenta)
        | Some annotation when Set.contains candidate annotation.focus && Set.contains candidate candidates -> 
            ColouredString(candidate.ToString(), ConsoleColor.Yellow)
        | Some annotation when annotation.primaryHintHouse -> 
            if Set.contains candidate candidates then ColouredString(candidate.ToString(), ConsoleColor.DarkGreen)
            else CChar ' '
        | Some annotation when annotation.secondaryHintHouse -> 
            if Set.contains candidate candidates then ColouredString(candidate.ToString(), ConsoleColor.Green)
            else CChar ' '
        | _ -> 
            if Set.contains candidate candidates then CStr(candidate.ToString())
            else CChar ' '

File: src/console.ml
open Sudoku
open Format
open Hint
(*F# open FSharp.Compatibility.OCaml F#*)

let drawDigitCellContents (given : digit option) (current : cellContents) : consoleChar = 
    match given, current with
    | Some s, _ -> ColouredDigit(s, Blue, DefaultColour)
    | None, BigNumber s -> ColouredDigit(s, Red, DefaultColour)
    | None, PencilMarks _ -> CChar '.'

let drawDigitCellString (given : digit option) (current : cellContents) : consoleString =
    [drawDigitCellContents given current]

let drawBigNumber (annotation : annotation) (digit : digit) : consoleChar =
    if annotation.primaryHintHouse then
        match annotation.given with
        | Some _ -> ColouredDigit(digit, Cyan, DefaultColour)
        | None -> ColouredDigit(digit, Yellow, DefaultColour)
    else if annotation.secondaryHintHouse then
        match annotation.given with
        | Some _ -> ColouredDigit(digit, DefaultColour, Blue)
        | None -> ColouredDigit(digit, DefaultColour, Red)
    else
        match annotation.given with
        | Some _ -> ColouredDigit(digit, Blue, White)
        | None -> ColouredDigit(digit, Red, DefaultColour)

let drawPencilMarks (annotation : Hint.annotation) (candidate : digit) (candidates : digits) : consoleChar =
    match annotation.setValue with
    | Some vv when vv = candidate -> 
        ColouredDigit(candidate, Red, DefaultColour)
    | Some _ when Digits.contains candidate candidates -> 
        ColouredDigit(candidate, Green, DefaultColour)
    | _ ->
        (match annotation.setValueReduction with
         | Some svr when svr = candidate && Digits.contains candidate candidates -> 
            ColouredDigit(candidate, Green, DefaultColour)
         | _ ->
            (if Digits.contains candidate annotation.reductions then
                ColouredDigit(candidate, Green, DefaultColour)
             else if Digits.contains candidate annotation.pointers then
                ColouredDigit(candidate, Magenta, DefaultColour)
             else if Digits.contains candidate annotation.focus && Digits.contains candidate candidates then
                ColouredDigit(candidate, Yellow, DefaultColour)
             else if annotation.primaryHintHouse then
                if Digits.contains candidate candidates then ColouredDigit(candidate, Cyan, DefaultColour)
                else CChar ' '
             else if annotation.secondaryHintHouse then
                if Digits.contains candidate candidates then ColouredDigit(candidate, Green, DefaultColour)
                else CChar ' '
             else
                if Digits.contains candidate candidates then ColouredDigit(candidate, Green, DefaultColour)
                else CChar ' '))

let drawDigitCellContentAnnotations centreCandidate (annotations : (cell * Hint.annotation) list) (cell : cell) (candidate : digit) : consoleChar = 

    let annotation = Smap.get Cell.comparer cell annotations in

    match annotation.current with
    | BigNumber s when centreCandidate = candidate -> drawBigNumber annotation s
    | BigNumber _ -> ColouredString(" ", Blue, White)
    | PencilMarks digits -> drawPencilMarks annotation candidate digits

let drawDigitCellContentAnnotationString (centreCandidate : digit) (annotations : (cell * Hint.annotation) list) (cell : cell) (candidate : digit) : consoleString =
    [drawDigitCellContentAnnotations centreCandidate annotations cell candidate]
